oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticSuggestIterator.java

```diff
@@ -0,0 +1,116 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.PriorityQueue;
+
+/**
+ * This class is in charge to extract suggestions for a given query. Suggestion is more like
+ * a completion result.
+ */
+class ElasticSuggestIterator implements Iterator<FulltextResultRow> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ElasticSuggestIterator.class);
+    protected final static String SUGGEST_PREFIX = "suggest?term=";
+
+    private final ElasticIndexNode indexNode;
+    private final ElasticRequestHandler requestHandler;
+    private final ElasticResponseHandler responseHandler;
+    private final String suggestQuery;
+
+    private Iterator<FulltextResultRow> internalIterator;
+    private boolean loaded = false;
+
+    ElasticSuggestIterator(@NotNull ElasticIndexNode indexNode,
+                           @NotNull ElasticRequestHandler requestHandler,
+                           @NotNull ElasticResponseHandler responseHandler) {
+        this.indexNode = indexNode;
+        this.requestHandler = requestHandler;
+        this.responseHandler = responseHandler;
+        this.suggestQuery = requestHandler.getPropertyRestrictionQuery().replace(SUGGEST_PREFIX, "");
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (!loaded) {
+            try {
+                loadSuggestions();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            loaded = true;
+        }
+        return internalIterator != null && internalIterator.hasNext();
+    }
+
+    @Override
+    public FulltextResultRow next() {
+        return internalIterator.next();
+    }
+
+    private void loadSuggestions() throws IOException {
+        BoolQueryBuilder suggestionQuery = requestHandler.suggestionMatchQuery(suggestQuery);
+        final SearchSourceBuilder searchSourceBuilder = SearchSourceBuilder.searchSource()
+                .query(suggestionQuery)
+                .size(100)
+                .fetchSource(FieldNames.PATH, null);
+        final SearchRequest searchRequest = new SearchRequest(indexNode.getDefinition().getRemoteIndexAlias())
+                .source(searchSourceBuilder);
+        SearchResponse res = indexNode.getConnection().getClient().search(searchRequest, RequestOptions.DEFAULT);
+        ArrayList<FulltextResultRow> results = new ArrayList<>();
+        PriorityQueue<ElasticSuggestion> pr = new PriorityQueue<>((a, b) -> Float.compare(b.score, a.score));
+        for (SearchHit doc : res.getHits()) {
+            if (responseHandler.isAccessible(responseHandler.getPath(doc))) {
+                for (SearchHit suggestion : doc.getInnerHits().get(":suggest").getHits()) {
+                    pr.add(new ElasticSuggestion(((List<String>) suggestion.getSourceAsMap().get("suggestion")).get(0), suggestion.getScore()));
+                }
+            }
+        }
+        for (ElasticSuggestion s : pr) {
```

Comment:
why do we need a `PriorityQueue` that needs to be re-processed? Shouldn't the suggestions be already ordered by score?

Reply:
Yes, suggestions are sorted but there can be multiple suggestions if we have two or more  suggest properties on a node.
e.g.
 /node1 --> suggest on title(score 1.9) and description(score 0.5)
/node2 -->  suggest on description(score 1.8) and title(score 1.5)
So suggestions are sorted in internally but combined together they are not.
In above case result should be
 /node1/title
/node2/description
/node2/title
/node2/description

Reply:
Got it now. Thanks.

I think we can optimize and simplify things a little bit. `ElasticSuggestion` can extend `FulltextResultRow`. In this way, you can directly return the `PriorityQueue` iterator avoiding the creation of the `ArrayList`. To make it work, you need also to change the `internalIterator` declaration to:

```java
private Iterator<? extends FulltextResultRow> internalIterator;
```

Reply:
Done.